Experiment - Playwright: save authentication cookies for common users prior to executing the test.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes an addition to the setup/teardown process to support the storing of authentication cookies for commonly used users with the goal of speeding up subsequent login attempts.

Key changes:
- new environment variable `SAVE_AUTH_COOKIES` to control behavior of new code path.
- creation of `globalSetup` and `globalTeardown` to save authentication cookies.
- additional branching in `LoginPage.login` method to handle the use of saved authentication cookie.

Background:

Since the migration to use Playwright framework for our e2e tests, we have occasionally seen issues such as #57440 occur intermittently. The frequency at which this occurred remained at an acceptable level until fairly recently when a moderate spike has been observed.

My working hypothesis has consistently been that it is likely due to a combination of the user hitting e2e-specific code path and becoming rate limited. To support this hypothesis, often failures of the `Log In` action is observed across multiple files and simultaneously across mobile and desktop variants.

Details:

This feature consists of the following parts:

#### Setup/Teardown
The `globalSetup` and `globalTeardown` are the first piece of the feature. It is controlled by an environment variable `SAVE_AUTH_COOKIES` and will only execute if set to `true`. Otherwise, the existence of this piece is transparent to Jest and the end user.

The `globalSetup` imports Playwright and a set of manually called methods to navigate to the Calypso BaseURL endpoint, fill in the credentials, submit and wait until the home dashboard loads. This is repeated for three of the mainly used users. At end of each iteration, a cookie file is written out to `test/e2e/cookies` matching the `accountType` of the user.

Importantly, an environment variable `COOKIE_PATH` is set here due to inability for Jest to share global variables set here with rest of the test runner process. See [documentation](https://jestjs.io/docs/configuration#globalsetup-string). 

#### Login
The method `LoginPage.login` normally accepts either an `accountType` string or `username/credential` string pair. The ability to accept `accountType` is used here to implement the new feature in a conditional, as recall that the file name of the authenticated cookie matches the `accountType`.

When the method is invoked, it checks whether an `accountType` forms part of the params. 
If true, it proceeds to look for a cookie file at the path `COOKIE_PATH` specifies.

If a cookie is found, the method will proceed to apply the cookie file to the BrowserContext.
Next, it will load the `/` endpoint.

If the cookie file was valid, by accessing `/` it should redirect user to the `/home/` endpoint. This is ensured via a `waitForNavigation` call.

In all steps within the `LoginPage.login` method an exception is raised relating to the authenticated cookie method, the exception is suppressed as a noop and the method continues to attempt a normal login (as with before).

The intention here is to not require the test developer to have to think about how to handle the two paths, or require  rewrite of existing tests.

#### Performance

We see a potentially large time savings:
![image](https://user-images.githubusercontent.com/6549265/139910151-356b264e-bc74-41c9-985b-4014dc07ee67.png)

The initial setup costs approx. 7-10 seconds to perform. 

Compare the times taken for the Log In action to complete:

Current trunk
![image](https://user-images.githubusercontent.com/6549265/139910776-60a69272-3f73-453e-aa0d-f967a7caaeb7.png)

Saved cookie
![image](https://user-images.githubusercontent.com/6549265/139910809-0d968cff-b177-4456-8159-1689a94143bb.png)

#### Testing instructions

- [ ] normal tests
  - [x] mobile
  - [x] desktop
  - [x] pre-release

Closes $57440.

worldomonation committed:
Add plumbing to save login states for users.
- new file `globalSetup.js` to launch an instance of Playwright Chromium, then log in to the list of user accounts defined in the method, saving cookies for the session under `/cookies`.
- add `globalSetup` to jest config.
- install Playwright 1.14 in `test/e2e/`.

worldomonation committed:
Add method to set cookies.

worldomonation committed:
Store the cookie directory as an environment variable from `globalSetup.js`, which will then be read by `environment.ts` to set a constant.

worldomonation committed:
Use COOKIES_PATH in BrowserManager.

worldomonation committed:
Add file to teardown
Convert global setup/teardown files to TypeScript.

worldomonation committed:
Modify LoginPage.login to work with both credentials for a hard login as well as a cookie-based 'soft' login.

worldomonation committed:
Use correct file reference.

worldomonation committed:
Enable saving of cookies

worldomonation committed:
Set user agent string.

worldomonation committed:
Revert unintended change to default.json.
Move `setLoginCookie` behind a try/catch so as to not fail for users where the cookie file was not generated.

worldomonation committed:
Remove unused import.
Correct typo in comment.

worldomonation commented:
@dpasque Pinging for a preview pass on this PR. 
What do you think about some of the assumptions I've made? (eg. no better way to share a global variable from `globalSetup`).

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -156,18 +160,43 @@ export class LoginPage {
 		// redirect of `<host>/home/<blogUrl>.
 		const landingUrl = options?.landingUrl ? options.landingUrl : `**/home/**`;
 
+		// If there is a stored cookie for the user, try that first.
+		if ( 'account' in credentials ) {
```
One thought I had was to split the `login` methods to two overloaded methods - one accepting `LoginCredentials` type and the other accepting a `TestAccount` type.

It's been ages since I had to use overloading so I went to look it up, and it seems TypeScript doesn't have a super intuitive overloading (a là Java 7).

As such I've restructured this method to perform both 1) cookie authentication attempt and 2) normal authentication.

dpasque replied:
>it seems TypeScript doesn't have a super intuitive overloading (a là Java 7)
Yeah, JS/TS overloading is basically just doing what you're doing here - you have to have one function that handles all the argument options!

I think this function looks great - it's very linear and clear!! 🎉

worldomonation commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,47 @@
+import { mkdir } from 'fs/promises';
+import path from 'path';
+import config from 'config';
+import { chromium } from 'playwright';
```
I looked at importing `@automattic/calypso-e2e`, however the [jest-conditionals.ts](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/jest-conditionals.ts) file triggers the error `Do not import `@jest/globals` outside of the Jest test environment`.

scinos commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -484,6 +484,7 @@ fun playwrightPrBuildType( targetDevice: String, buildUuid: String ): BuildType
 					export PLAYWRIGHT_BROWSERS_PATH=0
 					export TEAMCITY_VERSION=2021
 					export HEADLESS=true
+					export SAVE_AUTH_COOKIES=true
```
Do we want to expose this as a TeamCity env variable so we can quickly turn it off? I think that can help diagnose future problems.

worldomonation replied:
Sounds like a good idea!

scinos commented on packages/calypso-e2e/src/environment.ts:
```diff
@@ -0,0 +1,2 @@
+export const COOKIES_PATH =
+	process.env.COOKIES_PATH !== null ? ( process.env.COOKIES_PATH as string ) : '';
```
I think this can be simplified by using the nullish operator
```suggestion
export const COOKIES_PATH = process.env.COOKIES_PATH ?? '';
```

scinos commented:
Very nice!! A couple of notes:

- What do you think about expanding this idea and storing `localStore` (or even the full user data dir!) as well? A quick search in Playwright docs suggests this could help: https://playwright.dev/docs/api/class-browsercontext/#browser-context-storage-state
- I haven't thought about it thoroughly, but a potential improvement would be to bake in the cookies in the Docker image itself, so we don't have to re-create them on every run, saving us some more time. Would that be possible?

scinos commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,47 @@
+	const browser = await chromium.launch();
+	const calypsoBaseURL = config.get( 'calypsoBaseURL' );
+	const cookieBasePath = path.join( process.cwd(), 'cookies' );
+	process.env.COOKIES_PATH = cookieBasePath;
```
This is ugly, but I don't know a better way to pass info to each test process. Maybe add a comment explaining why we need this here?

worldomonation replied:
Comments coming up. I agree that this is not the preferred approach - the separation of `globalXYZ` from rest of the test runner process is an understandable decision by the Jest team, but not something I agree with.

scinos commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,47 @@
+
+	const browser = await chromium.launch();
+	const calypsoBaseURL = config.get( 'calypsoBaseURL' );
+	const cookieBasePath = path.join( process.cwd(), 'cookies' );
```
Nit: I'd use something based on `__dirname` instead of `process.cwd()` to get the path. Using the cwd is a bit brittle if the user is running it from an IDE or using something like `yarn workspace run`

worldomonation replied:
Noted and will be revised in next commit.

worldomonation commented:
> Very nice!! A couple of notes:
> 
>     * What do you think about expanding this idea and storing `localStore` (or even the full user data dir!) as well? A quick search in Playwright docs suggests this could help: https://playwright.dev/docs/api/class-browsercontext/#browser-context-storage-state

This change currently uses `browserContext.storageState` to store the cookies.
Looking at the resulting file only the `cookie` portion is populated. 

The full user data directory is an interesting thought. I assume you're referring to saving the entire user data directory [on disk](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/user_data_dir.md#Linux)?

>     * I haven't thought about it thoroughly, but a potential improvement would be to bake in the cookies in the Docker image itself, so we don't have to re-create them on every run, saving us some more time. Would that be possible?

I think this is a logical extension of the concept in this PR. From what I can tell, the login cookies are valid for ~3 days, and I recall the `ci-e2e` image is regenerated once a day? It would eliminate a further ~7-10 seconds from each task and seeing how often the `ci-e2e` image is used it will quickly add up.

I'll spin up a new task to explore this idea in the coming days.

worldomonation committed:
PR review comments:
- expose SAVE_AUTH_COOKIES env var to TeamCity.
- use nullish coalescing operator.
- extensive comments outlining why env var must be used to communiate from globalSetup.

scinos commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -459,6 +459,10 @@ fun playwrightPrBuildType( targetDevice: String, buildUuid: String ): BuildType
 			cleanCheckout = true
 		}
 
+		params {
+			param("env.SAVE_AUTH_COOKIES", "true")
+		}
```
Nit: you can also use a `checkbox` instead of asking the user to type "true" or "false" (that being said, we have a ton of parameters that could be checkboxes but aren't, so this is not a new problem. Feel free to ignore, we can change this one if/when we change the rest of the boolean parameters.

```suggestion
		params {
			checkbox(
				name = "env.SAVE_AUTH_COOKIES",
				value = "true",
				label = "Save authentication cookies",
				description = "Login once and reuse auth cookies for all specs",
				checked = "true",
				unchecked = "false"
			)
		}
```

worldomonation replied:
Neat, I did not know about this!

worldomonation replied:
Fixed in 8d1fe0d0de45743c639026ebc118f29237f5f07c.

dpasque commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,65 @@
+	// its presence when running each test file.
+	process.env.COOKIES_PATH = cookieBasePath;
+
+	const userList = [ 'gutenbergSimpleSiteUser', 'wooCommerceUser', 'defaultUser' ];
```
Do we want the new `eCommerceUser` here instead of the woo one?

worldomonation replied:
I'll rebase and switch over the user. Thanks for catching!

worldomonation replied:
Ah shoot, forgot to do this in the shuffle of things. Will address in next patch.

dpasque commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,65 @@
+
+	const userList = [ 'gutenbergSimpleSiteUser', 'wooCommerceUser', 'defaultUser' ];
+
+	for await ( const user of userList ) {
```
Small nit, but you don't need the `await` here! The `await` in the for line is just if your iterable is itself async (think of using a for loop over a data stream from a web call or something like that)

worldomonation replied:
Ah shoot, forgot to do this in the shuffle of things. Will address in next patch.

dpasque commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,65 @@
+	// The cookie file will be stored in the `test/e2e/cookies` directory.
+	const cookieBasePath = path.join( __dirname, '../../', 'cookies' );
+	await mkdir( cookieBasePath, { recursive: true } );
+	// This is important!
+	// Jest does not permit values created in globalSetup/globalTeardown to be read
+	// by the test suites themselves. See https://jestjs.io/docs/configuration#globalsetup-string.
+	// As such the only way to communicate the cookie path to the test suites is to
+	// set this environment variable and have the `@automattic/calypso-e2e` module check for
+	// its presence when running each test file.
+	process.env.COOKIES_PATH = cookieBasePath;
```
Great comment! I agree - I think this is our best option here.

worldomonation committed:
Use rich checkbox feature for environment variable flag.

dpasque commented on test/e2e/lib/jest/globalSetup.ts:
```diff
@@ -0,0 +1,65 @@
+import { chromium } from 'playwright';
+
+// Save cookies for defaultUser.
+export default async (): Promise< void > => {
```
Alright, I want to preface this feedback as that I realize it's asking a lot for potentially a small change, so we don't definitely need to do this, but I'll make my pitch 🙂 

The short is: it feels weird for me to not have the bulk of this setup function be in the `calypso-e2e` package. What I had imagined was a function in `calypso-e2e` with a signature like...

```typescript
cacheLoginCookiesForAccounts( accounts: string[], cookieBasePath: string): Promise<void>
```

And then this global setup function would call that function and pass in the account names and the path to use for the cookies.

As I see it, there a couple drawbacks from not having the core functionality in `calypso-e2e`:
1. We're having to re-implement some functionality (like logging in, spinning up a browser context, etc) that is already implemented in the `calypso-e2e` package. It would be best if we could re-use what we've already implemented!
2. We're forced to split implementation details across two packages, and this ends up breaking encapsulation. For example, it feels weird in this file that we have to know the specific cookie ENV variable name we need to set for the `calypso-e2e` package to use later, and to know the naming scheme we need to use for saving off cookies (to later be consumed again by a function in the `calypso-e2e` package). I think that might lead to fragility over time, and it would make it harder for future consumers of the `calypso-e2e` package -- they effectively have to read the underlying code to use the library.

But, as you mention, the big blocker here is that `calypso-e2e` uses `@jest/globals`, which got me thinking... Why does it?

I think it would be powerful if we could completely decouple the `calypso-e2e` package from `jest` - this means that anyone could use this project and then use whatever test runner they want! It also means that if we want to switch runners in the future (to the Playwright runner, e.g.), we wouldn't have to touch the `calypso-e2e` package at all, making the change even easier!

To do that decoupling, we only have to do a couple relatively small things:
1. Move the `jest-conditionals.ts` file to `test/e2e/lib` - it's coupled to jest, so I think it makes sense to live with the rest of jest stuff
2. Change `setupHooks` so that it follows more of a dependency injection pattern. There's a couple different ways we could do this: we could have `setupHooks` accept the before/after functions. Or alternatively, we could have `calypso-e2e` just export functions that do the bulk of what is called during before/after, and then we could have a local `setupHooks` function in `test/e2e/lib` that calls those functions within the jest `beforeAll` and `afterAll` functions.


Anyway, I realize that's a lot, and maybe way too much for here, but wanted to throw the idea and discussion out there!

dpasque reviewed (approved):
Approving with some small suggestions, and a bigger discussion point!

I think the approach you have here works great and will be a great boon to test performance! I do layout a discussion point for an alternative way to architect this, but it's by no means required. Let me know what you think, and feel free to ping me on Slack for further discussion!
